[Concurrency] Basic support for actor classes and actor isolation

DougGregor opened the pull request:
Introduce the "actor class" syntax. Ensure that it is only used for
root classes or classes that inherit from other actor classes. Within
actors, implement a basic set of isolation rules for actors, which includes:

* Asynchronous actor methods can be invoked on any actor instance.
* All mutable instance properties and synchronous instance methods are
only accessible via "self" or "super" within the context of the actor.
* Closures defined within an actor method are considered to be a
context distinct from the actor itself, are therefore are subject to
the same restrictions as above.

Together with a scheduler that ensures that asynchronous invocations
for a particular actor are serialized (at least up until they hit a
suspension point or return), this serves to isolate the actor's
instance state in a shallow manner: references to shared mutable state
can still be passed among actors, code can access global variables,
and one can still form and pass around unsafe pointers.

Pair this with some basic checking within actor contexts that
identifies a few obvious potential kinds of race conditions:

* References to captured mutable local variables from within closures.
* References to mutable global and static/class variables.

The latter are warnings, for which we currently don't have a great
suppression mechanism. That, and a lot of tuning, are still to come.

DougGregor committed:
[Concurrency] Add parsing support for actor classes.

Introduce the "actor class" syntax. Ensure that it is only used for
root classes or classes that inherit from other actor classes.

DougGregor committed:
[Concurrency] Implement basic actor isolation rules.

Implement a basic set of isolation rules for actors, which includes:

* Asynchronous actor methods can be invoked on any actor instance.
* All mutable instance properties and synchronous instance methods are
only accessible via "self" or "super" within the context of the actor.
* Closures defined within an actor method are considered to be a
context distinct from the actor itself, are therefore are subject to
the same restrictions as above.

Together with a scheduler that ensures that asynchronous invocations
for a particular actor are serialized (at least up until they hit a
suspension point or return), this serves to isolate the actor's
instance state in a shallow manner: references to shared mutable state
can still be passed among actors, code can access global variables,
and one can still form and pass around unsafe pointers.

Pair this with some basic checking within actor contexts that
identifies a few obvious potential kinds of race conditions:

* References to captured mutable local variables from within closures.
* References to mutable global and static/class variables.

The latter are warnings, for which we currently don't have a great
suppression mechanism. That, and a lot of tuning, are still to come.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

atrick commented on test/Concurrency/actor_isolation.swift:
```diff
@@ -0,0 +1,109 @@
+    // Closures.
+    let localConstant = 17
+    var localVar = 17 // expected-note 2{{var declared here}}
+    acceptClosure {
```
The `acceptClosure` errors are way more restrictive than I thought. In the proposal, a function type must be declared `@concurrent` if it potentially crosses an actor boundary.

ktoso replied:
This surprised me as well, are not just escaping closures troublesome? To schedule a closure off somewhere I'd have to escape it, right?

DougGregor replied:
The idea is to assume that non escaping functions won't be executed concurrently but escaping functions would be. I've now implemented that here. All of these rules are subject to change; this is very much an experiment.

atrick commented on test/Concurrency/actor_isolation.swift:
```diff
@@ -0,0 +1,109 @@
+      _ = localConstant
+    }
+
+    acceptEscapingClosure {
```
It's strange that `self.synchronous` is left out of the `acceptEscapingClosure` cases.

DougGregor replied:
Error of omission, I'll fix before merging. Thanks!

atrick reviewed (commented):
I looked at this for the sake of understanding the isolation rules. If the rules are written down anywhere else, I'd love to know.

ktoso reviewed (approved):
Fantastic stuff @DougGregor , minor comment/question inline :)

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -161,3 +162,457 @@ bool IsAsyncHandlerRequest::evaluate(
+  /// class that is isolated to the current actor instance.
+  ///
+  /// \returns the type of the actor.
+  static ClassDecl *getActorIsolatingInstanceMEmber(ValueDecl *value) {
```
Thanks! Fixed locally

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -161,3 +162,457 @@ bool IsAsyncHandlerRequest::evaluate(
+    /// The declaration can only be referenced from the given local context,
+    /// or a context that cannot execute concurrently with code in that
+    /// context.
+    Local,
```
It's not really about actors, though. This is about captured locals being accessed in code that can execute concurrently. I'll rename to `LocalCapture` to make that more clear.

ktoso replied:
Ah I see, yeah that reads well, thanks :)

DougGregor committed:
[Concurrency] Improve/fix some naming for actor isolation checking.

DougGregor commented on test/Concurrency/actor_isolation.swift:
```diff
@@ -0,0 +1,109 @@
+    _ = otherActor.text[0] // expected-error{{actor-isolated property 'text' can only be referenced on 'self'}}
+
+    // Global data is okay if it is immutable.
+    _ = immutableGlobal
```
There is a giant gaping hole here around reference types that I'm not going to touch in this particular pull request. We won't get away without adding type-level annotations, which need to be separately considered and introduced.

ktoso replied:
Right, got it :)

DougGregor committed:
[Concurrency] Treat actor-isolated references from closures as errors.

Following the approach that actors are meant to be isolated, promote
the warning about accessing actor-isolated state from a
potentially-concurrent context to an error.

DougGregor committed:
[Concurrency] Assume non-escaping closures don't run concurrently.

Non-escaping closures are assumed to run non-concurrently, so don't
complain about accesses to actor-isolated state on self within them.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci smoke test

DougGregor commented:
Windows failure looks spurious:
```
fatal error C1083: Cannot open compiler generated file: 'T:\lldb\source\Plugins\Language\ObjC\CMakeFiles\lldbPluginObjCLanguage.dir\NSIndexPath.cpp.obj': Permission denied
```

davezarzycki commented:
If actors cannot inherit from classes and vice-versa, shouldn't they have a separate type within the type system? Or is this just implementation expedience?
